src/mame/olivetti/att6300p.cpp

```diff
@@ -0,0 +1,1050 @@
+void att6300p_state::dma_dack_w(int state)
+{
+	m_isabus->dack_line_w(Channel, state);
+
+	if (!state)
+	{
+		m_dma_active |= 1 << Channel;
+		if (Channel == 0)
+			m_dmac->dreq0_w(0);
+		if (m_tc)
+			m_isabus->eop_w(Channel, ASSERT_LINE);
+	}
+	else
+	{
+		m_dma_active &= ~(1 << Channel);
+		if (m_tc)
+			m_isabus->eop_w(Channel, CLEAR_LINE);
+	}
+}
+
+void att6300p_state::dma_tc_w(int state)
+{
+	m_tc = (state == ASSERT_LINE);
+	for (int channel = 0; channel < 4; channel++)
+		if (BIT(m_dma_active, channel))
+			m_isabus->eop_w(channel, state);
+}
+
+void att6300p_state::dreq0_ck_w(int state)
+{
+	if (state && !m_dreq0_ck && !BIT(m_dma_active, 0))
+		m_dmac->dreq0_w(1);
+
+	m_dreq0_ck = state;
+}
+
+void att6300p_state::speaker_ck_w(int state)
+{
+	if (state)
+		m_ctrlport_b |= 0x20;
+	else
+		m_ctrlport_b &= 0xdf;
+
+	update_speaker();
+}
+
+void att6300p_state::update_speaker()
+{
+	if (BIT(m_ctrlport_a, 1) && BIT(m_ctrlport_b, 5))
+	{
+		m_speaker->level_w(1);
+		m_ctrlport_b &= 0xef;
+	}
+	else
+	{
+		m_speaker->level_w(0);
+		m_ctrlport_b |= 0x10;
+	}
+}
+
+void att6300p_state::update_ir1()
+{
+    // TimeSlice interrupt is shared with KBD interrupt
+	m_pic->ir1_w((m_kb_int || m_ts_int) ? 1 : 0);
+}
+
+int att6300p_state::kbc_t0_r()
+{
+	return 1;
+}
+
+int att6300p_state::kbc_t1_r()
+{
+	return BIT(m_ctrlport_a, 7);	// Keyboard Enable
+}
+
+uint8_t att6300p_state::keyboard_data_r()
+{
+	m_kb_int = false;
+	update_ir1();
+	return m_kbc->upi41_master_r(0);
+}
+
+uint8_t att6300p_state::keyboard_status_r()
+{
+	return m_kbc->upi41_master_r(1);
+}
+
+void att6300p_state::ctrlport_a_w(uint8_t data)
+{
+	const bool spkrdata_en_dis = BIT(data ^ m_ctrlport_a, 1);
+	const bool iochk_en_dis = BIT(data ^ m_ctrlport_a, 4);
+
+	m_pit->write_gate2(BIT(data, 0));
+
+	if (BIT(m_ctrlport_a, 4) && !m_chck_active)
+		m_ctrlport_b &= 0xbf;
+
+	m_ctrlport_a = data;
+
+	if (spkrdata_en_dis)
+		update_speaker();
+	if (iochk_en_dis)
+		update_nmi();
+}
+
+uint8_t att6300p_state::ctrlport_a_r()
+{
+	return m_ctrlport_a;
+}
+
+uint8_t att6300p_state::ctrlport_b_r()
+{
+	// Bit 0 = NC
+	// Bit 1 = DSW2-5 (80287 present)
+	// Bit 2 = ~RI1
+	// Bit 3 = ~DSR1
+	// Bit 4 = ~OUT2 (8254)
+	// Bit 5 = OUT2
+	// Bit 6 = IOCHK
+	// Bit 7 = MBMERR (MRD parity check)
+
+	if (BIT(m_dsw->read(), 12))
+		m_ctrlport_b |= 0x02;
+	else
+		m_ctrlport_b &= 0xfd;
+
+	return m_ctrlport_b;
+}
+
+void att6300p_state::chck_w(int state)
+{
+	m_chck_active = (state == 0);
+	if (m_chck_active)
+	{
+		if (!BIT(m_ctrlport_b, 6))
+		{
+			m_ctrlport_b |= 0x40;
+			update_nmi();
+		}
+	}
+	else if (BIT(m_ctrlport_a, 4))
+		m_ctrlport_b &= 0xbf;
+}
+
+void att6300p_state::nmi_enable_w(uint8_t data)
+{
+	m_nmi_enable = BIT(data, 7);
+	update_nmi();
+}
+
+void att6300p_state::update_nmi()
+{
+	m_nmi_active = (m_trapio_active ||
+	  (m_nmi_enable && BIT(m_ctrlport_b, 6) && !BIT(m_ctrlport_a, 4)));
+
+	if (m_nmi_active)
+		m_maincpu->set_input_line(INPUT_LINE_NMI, ASSERT_LINE);
```

Comment:
Are you sure this isn't pulsed? I. e. something like `m_maincpu->pulse_input_line(INPUT_LINE_NMI, attotime::zero);`.